### PR TITLE
add `taplo` CI + format all TOML files

### LINKED ISSUE #914: formatting on TOML files
our TOML files have random formatting standards

we should have some CI to make sure all TOML files are properly formatted under the same standard, just like we do with `rustfmt`

### PR DESCRIPTION
**Pull request by plebhash**
close #914 by:
- adding a CI step based on [`taplo`](https://taplo.tamasfe.dev/)
- applying to all TOML files

### CODE REVIEW

**Comment by jbesraa** [comment]
Thanks @plebhash. Could you please split the second commit(the actual formatting) into commit-for-each-crate?

**Comment by plebhash** [comment]
> Thanks @plebhash. Could you please split the second commit(the actual formatting) into commit-for-each-crate?

done

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:11]: Personally, I like the space between the `[` and `]` in arrays because it is easier on my eyes, if you don't have a preference my suggestion is:

```suggestion
# Include white space padding in single-line arrays for better readability.
compact_arrays = false
```

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:8]:
```suggestion
# Collapse arrays that do not exceed the maximum column width and do not contain comments.
```

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:13]: Just like with `compact_arrays`, I like to have spacing between `{` and `}`. If you agree in setting `compact_arrays = true`, we should change this to true as well:

```suggestion
# Include white space padding in inline tables for better readability.
compact_inline_tables = true
```

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:16]: If you disagree with the above suggestion of setting this to 100 instead of 160, let's still update the comment (there is a minor typo there as well, "expagit nsion" -> "expansion". So if you do not merge any of my suggestions on this, be sure to fix that).

```suggestion
# Maximum column width in characters.
column_width = 160
```

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:23]:
```suggestion
# Do not reorder keys within tables; keep the original order.
```

File: taplo.toml
Comment (rrybarczyk) [inline taplo.toml:25]: Update comment for clarity of what setting is doing and ensure comments do not exceed 100 (if you keep 160 for max width, follow that instead).

```suggestion
# Limit the number of consecutive blank lines to 1 to reduce unnecessary whitespace. This does not
# affect the whitespace at the end of the document, as it is always stripped.
```

**Review by rrybarczyk — CHANGES_REQUESTED** [review]
Love the addition of `taplo`.

* Should we add instructions somewhere to install `talpo` with `cargo install taplo-cli`?
* Should we add instructions on running `taplo` to apply the format with `taplo fmt`?

File: taplo.toml
Comment (plebhash) [reply taplo.toml]: that would have been the opposite outcome

`compact_inline_tables = false` results in:
```
async-std           = { version = "1.10.0", features = [ "attributes" ] }
```

`compact_inline_tables = true` results in:
```
async-std           = {version = "1.10.0", features = [ "attributes" ]}
```

we don't need to change anything here

File: taplo.toml
Comment (plebhash) [reply taplo.toml:1]: from these suggestions, the only one I would prefer to diverge from is `align_entries = true`

**Comment by plebhash** [comment]
> * Should we add instructions somewhere to install `talpo` with `cargo install taplo-cli`?
> * Should we add instructions on running `taplo` to apply the format with `taplo fmt`?

I added some instructions https://github.com/stratum-mining/stratum/pull/915/commits/dcd231f7505219096617e2f73892954165afe603

**Review by rrybarczyk — APPROVED** [review]
I noticed that the `align = true` and the `sort = true` does not apply to all `Cargo.toml`s `[dependencies]` table (but appears to for all other tables), for example `roles/jd-server/Cargo.toml`. I have cloned this branch and ran `taplo` and tinkered with some settings, but I cannot figure out why these are not being universally applied. 

I think this is fine as is and we should merge. Just good to get this in the books and can always modify later.

**Comment by plebhash** [comment]
> I noticed that the `align = true` and the `sort = true` does not apply to all `Cargo.toml`s `[dependencies]` table (but appears to for all other tables), for example `roles/jd-server/Cargo.toml`. I have cloned this branch and ran `taplo` and tinkered with some settings, but I cannot figure out why these are not being universally applied.
> 
> I think this is fine as is and we should merge. Just good to get this in the books and can always modify later.

Here's the taplo docs on the options available: https://taplo.tamasfe.dev/configuration/formatter-options.html

The `align_entries` is described as:
> Align entries vertically. Entries that have table headers, comments, or blank lines between them are not aligned.

So it does seem to come with some caveats and edge cases.

I fed `roles/jd-server/Cargo.toml` to chatGPT and asked why `align_entries` wasn't being applied to `[dependencies]`, and here's the answer:

> The alignment is not being applied to the dependencies section due to the presence of entries with multiline values (lists and tables) that are not aligned by the align_entries option in Taplo. This option only aligns single-line key-value pairs. When multiline values are present, alignment is not performed to avoid disrupting the readability and structure of these values.

> Here’s your dependencies section with comments indicating where multiline values are causing the alignment to be skipped:

```
[dependencies]
stratum-common      = { version = "1.0.0", path = "../../common" }
async-channel       = "1.5.1"
async-recursion     = "0.3.2"
binary_sv2          = { version = "^1.0.0", path = "../../protocols/v2/binary-sv2/binary-sv2" }
buffer_sv2          = { version = "^1.0.0", path = "../../utils/buffer" }
codec_sv2           = { version = "^1.0.1", path = "../../protocols/v2/codec-sv2", features = [
    "noise_sv2",
    "with_buffer_pool",
] }  # Multiline value
framing_sv2         = { version = "^1.1.0", path = "../../protocols/v2/framing-sv2" }
network_helpers_sv2 = { version = "2.0.0", path = "../roles-utils/network-helpers", features = [
    "with_tokio",
    "with_buffer_pool",
] }  # Multiline value
roles_logic_sv2     = { version = "^1.0.0", path = "../../protocols/v2/roles-logic-sv2" }
serde               = { version = "1.0.89", default-features = false, features = [ "derive", "alloc" ] }  # Multiline value
futures             = "0.3.25"
tokio               = { version = "1", features = [ "full" ] }  # Multiline value
toml                = { version = "0.5.6", git = "https://github.com/diondokter/toml-rs", default-features = false, rev = "c4161aa" }  # Multiline value
tracing             = { version = "0.1" }
tracing-subscriber  = { version = "0.3" }
error_handling      = { version = "1.0.0", path = "../../utils/error-handling" }
nohash-hasher       = "0.2.0"
key-utils           = { version = "^1.0.0", path = "../../utils/key-utils" }
```

> In this case, entries like codec_sv2, network_helpers_sv2, serde, tokio, and toml contain multiline values (e.g., arrays or tables), preventing the align_entries option from aligning the entire dependencies section.

---

with regards to the `sort` option, I couldn't find it anywhere in the docs so I'm actually wondering if that is really a feature that `taplo` provides?

@rrybarczyk I see you suggested its introduction [here](https://github.com/stratum-mining/stratum/pull/915#discussion_r1650148061), do you recall what you took as reference for that suggestion?

also note that `taplo` will not complain if you feed it an unexistant option... for example, having `ssssort = true` inside `.taplo.toml` doesn't elicit any errors when I do `taplo fmt`, so it might be the case that `sort = true` is simply being ignored because it's not a real option

in case we confirm that `sort` is not a real option for `taplo`, we should probably remove this line since it's not really doing anything

**Review by jbesraa — COMMENTED** [review]
I am not really convinced about formatting `Cargo.toml` files. As long as they are not huge, the normal formatting is just fine. whats the motivation for including them here?

**Comment by plebhash** [comment]
> I am not really convinced about formatting `Cargo.toml` files. As long as they are not huge, the normal formatting is just fine. whats the motivation for including them here?

original motivation was just to bring some uniformity to them... there's random patterns accross multiple files

**Comment by jbesraa** [comment]
> > I am not really convinced about formatting `Cargo.toml` files. As long as they are not huge, the normal formatting is just fine. whats the motivation for including them here?
> 
> original motivation was just to bring some uniformity to them... there's random patterns accross multiple files
> 
> but also, if we include taplo on CI, we need all TOML files to conform to those rules, otherwise CI will break for every PR

I think if you are eager to format `Cargo.toml` files, the formatting should at least follow the original formatting. i.e.,  similar to what `cargo` produces when creating new crate. 

Cant `taplo` exclude files?

**Comment by plebhash** [comment]
> > > I am not really convinced about formatting `Cargo.toml` files. As long as they are not huge, the normal formatting is just fine. whats the motivation for including them here?
> > 
> > 
> > original motivation was just to bring some uniformity to them... there's random patterns accross multiple files
> > but also, if we include taplo on CI, we need all TOML files to conform to those rules, otherwise CI will break for every PR
> 
> I think if you are eager to format `Cargo.toml` files, the formatting should at least follow the original formatting. i.e., similar to what `cargo` produces when creating new crate.

When I first started this PR, I introduced a `.taplo.toml` file without giving much thought.
Later on the first round of reviews @rrybarczyk made some suggestions on this config file.

I assume that using `taplo fmt` without any config file would resort to whatever `cargo` defaults are, but I haven't validated this in practice. 

tbh I'm fine with this simplified approach as well, I don't really care for one particular flavor of formatting... my overall aim is to just have **some** standard rather than none, which leads to random formatting and gives room for sloppy PRs.

@rrybarczyk how do you feel about this?

> Cant `taplo` exclude files?

from my research on CLI args, no... but it is able to selectively format files if you feed it via stdin.

for example, if you only do `taplo fmt`, it will recursively format all `.toml` files on that directory/subdirectories.

but if you do `taplo fmt config.toml`, it will ignore all other files on that directory/subdirectories and focus specifically on `config.toml`

**Comment by rrybarczyk** [comment]
@plebhash my bad, I don't know where I got `sort` from, so yes let's remove since it DNE!

**Comment by rrybarczyk** [comment]
> > > > I am not really convinced about formatting `Cargo.toml` files. As long as they are not huge, the normal formatting is just fine. whats the motivation for including them here?
> > > 
> > > 
> > > original motivation was just to bring some uniformity to them... there's random patterns accross multiple files
> > > but also, if we include taplo on CI, we need all TOML files to conform to those rules, otherwise CI will break for every PR
> > 
> > 
> > I think if you are eager to format `Cargo.toml` files, the formatting should at least follow the original formatting. i.e., similar to what `cargo` produces when creating new crate.
> 
> When I first started this PR, I introduced a `.taplo.toml` file without giving much thought. Later on the first round of reviews @rrybarczyk made some suggestions on this config file.
> 
> I assume that using `taplo fmt` without any config file would resort to whatever `cargo` defaults are, but I haven't validated this in practice.
> 
> tbh I'm fine with this simplified approach as well, I don't really care for one particular flavor of formatting... my overall aim is to just have **some** standard rather than none, which leads to random formatting and gives room for sloppy PRs.
> 
> @rrybarczyk how do you feel about this?
> 
> > Cant `taplo` exclude files?
> 
> from my research on CLI args, no... but it is able to selectively format files if you feed it via stdin.
> 
> for example, if you only do `taplo fmt`, it will recursively format all `.toml` files on that directory/subdirectories.
> 
> but if you do `taplo fmt config.toml`, it will ignore all other files on that directory/subdirectories and focus specifically on `config.toml`

Yes I am good with the simplified approach!

**Comment by plebhash** [comment]
@jbesraa I removed `.taplo.toml` with the opinionated formatting from the PR

now, we are just doing the default formatting

**Comment by jbesraa** [comment]
> @jbesraa I removed `.taplo.toml` with the opinionated formatting from the PR
> 
> now, we are just doing the default formatting

You can exclude `Cargo.toml` from `taplo`, see this https://taplo.tamasfe.dev/configuration/file.html#exclude. 
There example is exactly what we want: you specify which folder you want to include in the `include` prop and in `exclude` you ignore files from the `included` stuff.

**Comment by plebhash** [comment]
> My main concern here is that we are using the same CI process + same formatting setup for a config intended for users and a config(Cargo.toml) we use, requirements between those two could diverge easily

I rebased. Now we have three separate `taplo` config files, each applying to specific categories of TOML files.

They don't have any specific formatting policies however, only `inlcude`/`exclude` filters.

@jbesraa do you have any specific ideas on formatting policies, or did you just want to make this PR "future-proof" in that sense?

File: protocols/v1/Cargo.toml
Comment (plebhash) [reply protocols/v1/Cargo.toml:18]: I'm not sure, the `.taplo*.toml` files are all empty with regards to actual formatting policies, they only do filtering for which file they will format

so I assume this is coming from the default formatting policy that `taplo` brings

**Comment by plebhash** [comment]
reverting this back to Draft status

merging this will have an undesired impact on our semver checking on CI, see this comment for more details: https://github.com/stratum-mining/stratum/pull/985#issuecomment-2181652506

we should get https://github.com/stratum-mining/stratum/pull/985 sorted out before merging this

@pavlenex we can remove this from milestone 1.0.2, since #985 could take a while and we don't want the next release blocked

**Comment by rrybarczyk** [comment]
Maybe someday.

**Pull request closed without merging**
